<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE TS>
<TS version="2.1" language="en_US">
<context>
    <name>QObject</name>
    <message>
        <source>SE reads or contigs</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>PE reads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>DIAMOND external tool support</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>The plugin supports DIAMOND: a sequence aligner for protein and translated DNA searches, designed for high performance analysis of big sequence data (https://github.com/bbuchfink/diamond)</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DiamondBuildTask</name>
    <message>
        <source>Build DIAMOND database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>There is no input files to build the database from</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DiamondClassifyTask</name>
    <message>
        <source>Classify sequences with DIAMOND</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Reads URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Paired reads URL is empty, but the &apos;paired reads&apos; option is set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>DIAMOND database URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>DIAMOND classification URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>URL to paired DIAMOND classification is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Taxon map URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Taxon nodes URL is empty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Invalid genetic code: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Unknown sensitivity value: %1, ignored.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::DiamondSupport</name>
    <message>
        <source>In general, DIAMOND is a sequence aligner for protein and translated DNA searches similar to the NCBI BLAST software tools. In UGENE it is integrated as one of the taxonomic classification tool.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::GenomesPreparationTask</name>
    <message>
        <source>Genomes preparation</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Genomes URLs are not set</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>File URL to write prepared genomes is empty</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::DiamondBuildPrompter</name>
    <message>
        <source>Use custom data to build %1 DIAMOND database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output DIAMOND database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>URL to the DIAMOND database file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>A name of the binary DIAMOND database file that should be created.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Genomic library</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Genomes that should be used to build the database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Build DIAMOND Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Build a DIAMOND formatted database from a FASTA input file.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::DiamondClassifyPrompter</name>
    <message>
        <source>Classify sequences from &lt;u&gt;%1&lt;/u&gt; with DIAMOND, use %2 database.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Input URL</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Input URL.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Input sequences</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>URL(s) to FASTQ or FASTA file(s) should be provided.

The input files may contain single-end reads, contigs, or &quot;left&quot; reads in case of the paired-end sequencing (see &quot;Input data&quot; parameter of the element).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>DIAMOND Classification</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>A list of sequence names with the associated taxonomy IDs, classified by DIAMOND.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Database</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Input a binary DIAMOND database file.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Genetic code</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Genetic code used for translation of query sequences (--query-gencode).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Sensitive mode</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>The sensitive modes (--sensitive, --more-sensitive) are generally recommended for aligning longer sequences. The default mode is mainly designed for short read alignment, i.e. finding significant matches of &gt;50 bits on 30-40aa fragments.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Frameshift</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Penalty for frameshift in DNA-vs-protein alignments. Values around 15 are reasonable for this parameter. Enabling this feature will have the aligner tolerate missing bases in DNA sequences and is most recommended for long, error-prone sequences like MinION reads.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Expected value</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Maximum expected value to report an alignment.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Matrix</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Scoring matrix (--matrix).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Gap open penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Gap open penalty (--gapopen).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Gap extension penalty</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Gap extension penalty (--gapextend).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Number of threads</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Number of CPU threads (--treads).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Block size</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Block size in billions of sequence letters to be processed at a time (--block-size). This is the main parameter for controlling the programâ€™s memory usage. Bigger numbers will increase the use of memory and temporary disk space, but also improve performance. The program can be expected to use roughly six times this number of memory (in GB).</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Index chunks</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>The number of chunks for processing the seed index (--index-chunks). This option can be additionally used to tune the performance. It is recommended to set this to 1 on a high memory server, which will increase performance and memory usage, but not the usage of temporary disk space.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Output file</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Specify the output file name.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Default</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Sensitive</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>More sensitive</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Skipped</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Classify Sequences with DIAMOND</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>In general, DIAMOND is a sequence aligner for protein and translated DNA searches similar to the NCBI BLAST software tools. However, it provides a speedup of BLAST ranging up to x20,000.&lt;br&gt;Using this workflow element one can use DIAMOND for taxonomic classification of short DNA reads and longer sequences such as contigs.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Taxonomy data is not set.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Taxonomy file &apos;%1&apos; is not found.</source>
        <translation type="unfinished"></translation>
    </message>
</context>
<context>
    <name>U2::LocalWorkflow::DiamondClassifyWorker</name>
    <message>
        <source>Not enough downstream reads datasets</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Not enough upstream reads datasets</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Taxonomy data is missed</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Cannot open classification report: %1</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Duplicate sequence name &apos;%1&apos; have been detected in the classification output.</source>
        <translation type="unfinished"></translation>
    </message>
    <message>
        <source>Broken Diamond report : %1</source>
        <translation type="unfinished"></translation>
    </message>
</context>
</TS>
